src/runtime/threading_backend.cc

```diff
@@ -0,0 +1,107 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file threading_backend.cc
+ * \brief Native threading backend
+ */
+#include <tvm/runtime/threading_backend.h>
+#include <dmlc/logging.h>
+#include <thread>
+#if defined(__linux__)
+#include <sched.h>
+#endif
+
+namespace tvm {
+namespace runtime {
+namespace threading {
+
+class NativeThreadGroup: public ThreadGroup {
+  public:
+    void Launch(std::vector<std::function<void()>> task_callbacks) {
+      for (int i = 0; i < task_callbacks.size(); ++i) {
+        threads_.emplace_back(task_callbacks[i]);
+      }
+      const char *val = getenv("TVM_BIND_THREADS");
+      if (val == nullptr || atoi(val) == 1) {
+        if (Size() <= std::thread::hardware_concurrency()) {
+          SetAffinity();
+        } else {
+          LOG(WARNING)
+            << "The thread affinity cannot be set when the number of workers is larger "
+            << "than the number of available cores in the system.";
+        }
+      }
+    }
+
+    inline void Join() {
+      for (std::thread& t : threads_) {
+        t.join();
+      }
+    }
+
+    // bind worker threads to disjoint cores
+    void SetAffinity() {
+#if defined(__ANDROID__)
+#ifndef CPU_SET
+      #define CPU_SETSIZE 1024
+      #define __NCPUBITS (8 * sizeof (uint64_t))
+      typedef struct {
+          uint64_t __bits[CPU_SETSIZE / __NCPUBITS];
+        } cpu_set_t;
+
+      #define CPU_SET(cpu, cpusetp) \
+        ((cpusetp)->__bits[(cpu)/__NCPUBITS] |= (1UL << ((cpu) % __NCPUBITS)))
+      #define CPU_ZERO(cpusetp) \
+        memset((cpusetp), 0, sizeof(cpu_set_t))
+#endif
+#endif
+        for (int i=0; i < threads_.size(); ++i) {
+#if defined(__linux__) || defined(__ANDROID__)
+              cpu_set_t cpuset;
+              CPU_ZERO(&cpuset);
+              CPU_SET(i, &cpuset);
+#if defined(__ANDROID__)
+              sched_setaffinity(threads_[i].native_handle(), sizeof(cpu_set_t), &cpuset);
+#else
+              pthread_setaffinity_np(threads_[i].native_handle(),
+                          sizeof(cpu_set_t), &cpuset);
+#endif
+#endif
+            }
+      }
+
+    size_t Size() { return threads_.size(); }
+
+  private:
+    std::vector<std::thread> threads_;
+};
+
+std::unique_ptr<ThreadGroup> CreateThreadGroup() {
```

Comment:
can be ThreadGroup::Create

Reply:
wouldn't that require a virtual static method (which isn't possible)?

Reply:
It does not have to be a virtual static method, it just have to be a static function and it can only be implemented once